### PR TITLE
Add kpatch actor

### PR DESCRIPTION
**Pull request by sm00th**
Add an actor that makes sure kpatch-dnf plugin and its config is carried
over to the container so that it installs appropriate kpatch-patch
during in-place-upgrade if needed.

### CODE REVIEW

File: repos/system_upgrade/common/actors/kpatch/actor.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from leapp.actors import Actor
+from leapp.libraries.stdlib import api
+from leapp.libraries.common.rpms import has_package
+from leapp.models import CopyFile, TargetUserSpacePreupgradeTasks, InstalledRedHatSignedRPM
+from leapp.tags import FactsPhaseTag, IPUWorkflowTag
+
+PLUGIN_PKGNAME = "kpatch-dnf"
+CONFIG_PATH = "/etc/dnf/plugins/kpatch.conf"
+
+class Kpatch(Actor):
```
Comment (pirat89) [inline repos/system_upgrade/common/actors/kpatch/actor.py:10]: please rename the actor (including dirname) to e.g. `CheckKpatch` (dir `checkkpatch`)  to make it more clear it's basically the check actor.

File: repos/system_upgrade/common/actors/kpatch/actor.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+from leapp.actors import Actor
+from leapp.libraries.stdlib import api
+from leapp.libraries.common.rpms import has_package
+from leapp.models import CopyFile, TargetUserSpacePreupgradeTasks, InstalledRedHatSignedRPM
+from leapp.tags import FactsPhaseTag, IPUWorkflowTag
```
Comment (pirat89) [inline repos/system_upgrade/common/actors/kpatch/actor.py:5]: move the actor to the `ChecksPhase` (the `ChecksPhaseTag`) - the actor is not getting any facts about the system, but does actions based on collected facts about the system.

File: repos/system_upgrade/common/actors/kpatch/actor.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+PLUGIN_PKGNAME = "kpatch-dnf"
+CONFIG_PATH = "/etc/dnf/plugins/kpatch.conf"
+
+class Kpatch(Actor):
+    """
+    Check is kpatch-dnf plugin is installed and if it is, install it and copy
+    over the config file so that the plugin can make a decision on whether any
+    kpatch-patch packages need to be installed during in-place upgrade.
+    """
```
Comment (pirat89) [inline repos/system_upgrade/common/actors/kpatch/actor.py:15]: Follow the suggested docstring format
```
"""
single line title

multiline body
"""
```

File: repos/system_upgrade/common/actors/kpatch/actor.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+    name = 'kpatch'
```
Comment (pirat89) [inline repos/system_upgrade/common/actors/kpatch/actor.py:17]: reflect the rename of the classname when it's done

**Review by pirat89 — CHANGES_REQUESTED** [review]
@sm00th on top of requested changes, move the actor under the `kernel` directory, so we have all kernel specific actors grouped (for easier orientation in number of existing actors)

**Comment by sm00th** [comment]
v2: addressed comments and flake8 codestyle warnings.

**Review by pirat89 — COMMENTED** [review]
@sm00th  It seems good to me. However, thinking about IPU 7 -> 8, is it valid? I mean, the official supported packager on RHEL 7 is YUM  only (if I am right, dnf is in extras and does not have to be configured in any way). So I am curious, whether the `CONFIG_PATH` is correct in case of RHEL 7, or whether it is expected something like
```
SRC_CONFIG_PATH = CONFIG_PATH if rhel7 else "/etc/yum/plugins/kpatch.conf"
```

**Comment by pirat89** [comment]
@sm00th pls make linter happy. You can use `make lint_fix` to fix the import automatically.

**Comment by sm00th** [comment]
> @sm00th It seems good to me. However, thinking about IPU 7 -> 8, is it valid? I mean, the official supported packager on RHEL 7 is YUM only (if I am right, dnf is in extras and does not have to be configured in any way).

This is a great question, thanks for noticing. It turns out we do not provide kpatch-dnf on rhel7 so it looks like we don't need (can't) support 7to8. Will the right course of action be to move this actor from `system_upgrade/common` to `system_upgrade/el8toel9/actors/kernel/checkkpatch/actor.py`?

**Comment by pirat89** [comment]
@sm00th it depends. is the yum plugin configuration compatible with the dnf plugin configuration? Does exist a way how to transform the original configuration to the new one for dnf if needed? if so, the actor can be still updated to support both upgrade paths. In case you want to support just 8 -> 9, it should be moved to the path you wrote.

**Comment by sm00th** [comment]
We don't have any yum alternative for the plugin so there is no config on rhel7. We will be only supporting 8to9 for now. O've already moved the actor in my last push.

**Review by pirat89 — APPROVED** [review]

File: repos/system_upgrade/el8toel9/actors/kernel/checkkpatch/actor.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+from leapp.actors import Actor
+from leapp.libraries.common.rpms import has_package
+from leapp.libraries.stdlib import api
+from leapp.models import CopyFile, InstalledRedHatSignedRPM, TargetUserSpacePreupgradeTasks
+from leapp.tags import CheckPhaseTag, IPUWorkflowTag
```
Comment (Rezney) [inline repos/system_upgrade/el8toel9/actors/kernel/checkkpatch/actor.py:5]:
```suggestion
from leapp.tags import ChecksPhaseTag, IPUWorkflowTag
```

File: repos/system_upgrade/el8toel9/actors/kernel/checkkpatch/actor.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    Check is kpatch-dnf plugin is installed and if it is, install it and copy
+    over the config file so that the plugin can make a decision on whether any
+    kpatch-patch packages need to be installed during in-place upgrade.
+    """
+
+    name = 'check_kpatch'
+    consumes = (InstalledRedHatSignedRPM,)
+    produces = (TargetUserSpacePreupgradeTasks,)
+    tags = (IPUWorkflowTag, CheckPhaseTag)
```
Comment (Rezney) [inline repos/system_upgrade/el8toel9/actors/kernel/checkkpatch/actor.py:23]:
```suggestion
    tags = (IPUWorkflowTag, ChecksPhaseTag)
```

**Review by Rezney — CHANGES_REQUESTED** [review]
```bash
2022-02-23 10:50:13.702 INFO     PID: 107934 leapp: Logging has been initialized
Process Process-101:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.6/site-packages/leapp/repository/actor_definition.py", line 30, in inspect_actor
    definition.load()
  File "/usr/lib/python3.6/site-packages/leapp/repository/actor_definition.py", line 179, in load
    self._module = importer.find_module(name).load_module(name)
  File "<frozen importlib._bootstrap_external>", line 399, in _check_name_wrapper
  File "<frozen importlib._bootstrap_external>", line 823, in load_module
  File "<frozen importlib._bootstrap_external>", line 682, in load_module
  File "<frozen importlib._bootstrap>", line 265, in _load_module_shim
  File "<frozen importlib._bootstrap>", line 684, in _load
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/etc/leapp/repos.d/system_upgrade/el8toel9/actors/kernel/checkkpatch/actor.py", line 5, in <module>
    from leapp.tags import CheckPhaseTag, IPUWorkflowTag
ImportError: cannot import name 'CheckPhaseTag'

Error: Inspection of actor in actors/kernel/checkkpatch failed
```

**Comment by sm00th** [comment]
Sorry about that, fixed.

**Comment by pirat89** [comment]
@sm00th one another question, have you tested it? Is there any kpatch* rpm on RHEL 9 that is providing fixes for kernel package? I am asking because I would like to see that the upgrade will not be broken when any such package for specific kernel release will be built. Currently I haven't seen anything around kpatch on RHEL 9 what's checking the `LEAPP_IPU_IN_PROGRESS` envar to be sure the kernel is not patched during the IPU. Maybe we should sync today/tomorrow via video call.

**Comment by sm00th** [comment]
> @sm00th one another question, have you tested it?
I tested the original version with FactsPhase.

> Is there any kpatch* rpm on RHEL 9 that is providing fixes for kernel package?

We don't have any real-life kpatch-patches for rhel9 yet, but we can create one for any kernel version for test purposes if/when needed.

> I am asking because I would like to see that the upgrade will not be broken when any such package for specific kernel release will be built. Currently I haven't seen anything around kpatch on RHEL 9 what's checking the `LEAPP_IPU_IN_PROGRESS` envar to be sure the kernel is not patched during the IPU.

We just merged the change today, I'll send you a link internally.

> Maybe we should sync today/tomorrow via video call.

We can discuss this over email.

**Comment by sm00th** [comment]
Anything else I can help with to get it merged in time?

**Comment by pirat89** [comment]
@sm00th Hi, I wrote you an email ~ 1h ago :) yes, we will need builds and I will appreciate if we sync today for the testing (builds for el9 will help!!) I want to be sure it works before we merge it.

**Comment by pirat89** [comment]
Tested with el8 where the active patching of kernel as enabled. Some snippets:
```
[snip]
-28 17:12:34.139 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:  kernel                                   x86_64  5.14.0-69.el9                            BASEOS                                                      561 k
2022-02-28 17:12:34.141 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:  kernel-core                              x86_64  5.14.0-69.el9                            BASEOS                                                       34 M
2022-02-28 17:12:34.142 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:  kernel-modules                           x86_64  5.14.0-69.el9                            BASEOS                                                       23 M
2022-02-28 17:12:34.144 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:  kpatch-patch-5_14_0-69                   x86_64  0-1.test.el9                             brew-task-repo-kpatch-patch-5_14_0-69-0-1.test.el9-scratch   12 k
2022-02-2

[snip]

2022-02-28 17:13:44.167 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:   Running scriptlet: subscription-manager-1.29.26-1.el9.x86_64         541/1313
2022-02-28 17:13:44.172 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:   Upgrading        : kpatch-dnf-0.4-2.el9.noarch                       542/1313
2022-02-28 17:13:44.177 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:   Running scriptlet: kpatch-dnf-0.4-2.el9.noarch                       542/1313
2022-02-28 17:13:44.179 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction: To enable automatic kpatch-patch subscription, run:
2022-02-28 17:13:44.180 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:    $ dnf kpatch auto
2022-02-28 17:13:44.181 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:
2022-02-28 17:13:44.196 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:   Upgrading        : kpatch-0.9.4-2.el9.noarch                         543/1313
2022-02-28 17:13:44.203 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:   Upgrading        : insights-client-3.1.7-1.el9.noarch                544/1313

[snip]

2022-02-28 17:13:45.216 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:   Installing       : kpatch-patch-5_14_0-69-0-1.test.el9.x86_64        580/1313
2022-02-28 17:13:45.304 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:   Running scriptlet: kpatch-patch-5_14_0-69-0-1.test.el9.x86_64        580/1313
2022-02-28 17:13:45.306 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction: installing /usr/lib/kpatch/5.14.0-69.el9.x86_64/kpatch-5_14_0-69-0-1_test.ko (5.14.0-69.el9.x86_64)
2022-02-28 17:13:45.308 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction: Created symlink /etc/systemd/system/multi-user.target.wants/kpatch.service → /usr/lib/systemd/system/kpatch.service.
2022-02-28 17:13:45.309 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction: chcon: can't apply partial context to unlabeled file '/var/lib/kpatch/5.14.0-69.el9.x86_64/kpatch-5_14_0-69-0-1_test.ko'
2022-02-28 17:13:45.310 DEBUG    PID: 724 leapp.workflow.RPMUpgrade.dnf_upgrade_transaction:
[snip]
```

**Review by pirat89 — APPROVED** [review]

**Pull request merged**